sdp-utils.c

```diff
@@ -0,0 +1,516 @@
+							success = FALSE;
+							break;
+						}
+						imported->o_name = g_strdup(name);
+						imported->o_addr = g_strdup(addr);
+						break;
+					}
+					case 's': {
+						imported->s_name = g_strdup(line+2);
+						break;
+					}
+					case 't': {
+						if(sscanf(line, "t=%"SCNu64" %"SCNu64, &imported->t_start, &imported->t_stop) != 2) {
+							if(error)
+								g_snprintf(error, errlen, "Invalid t= line: %s", line);
+							success = FALSE;
+							break;
+						}
+						break;
+					}
+					case 'c': {
+						char addrtype[6], addr[256];
+						if(sscanf(line, "c=IN %5s %255s", addrtype, addr) != 2) {
+							if(error)
+								g_snprintf(error, errlen, "Invalid c= line: %s", line);
+							success = FALSE;
+							break;
+						}
+						if(!strcasecmp(addrtype, "IP4"))
+							imported->c_ipv4 = TRUE;
+						else if(!strcasecmp(addrtype, "IP6"))
+							imported->c_ipv4 = FALSE;
+						else {
+							if(error)
+								g_snprintf(error, errlen, "Invalid c= line (unsupported protocol %s): %s", addrtype, line);
+							success = FALSE;
+							break;
+						}
+						imported->c_addr = g_strdup(addr);
+						break;
+					}
+					case 'a': {
+						janus_sdp_attribute *a = g_malloc0(sizeof(janus_sdp_attribute));
+						line += 2;
+						char *semicolon = strchr(line, ':');
+						if(semicolon == NULL) {
+							a->name = g_strdup(line);
+							a->value = NULL;
+						} else {
+							if(*(semicolon+1) == '\0') {
+								if(error)
+									g_snprintf(error, errlen, "Invalid a= line: %s", line);
+								success = FALSE;
+								break;
+							}
+							*semicolon = '\0';
+							a->name = g_strdup(line);
+							a->value = g_strdup(semicolon+1);
+							*semicolon = ':';
+						}
+						imported->attributes = g_list_append(imported->attributes, a);
+						break;
+					}
+					case 'm': {
+						janus_sdp_mline *m = g_malloc0(sizeof(janus_sdp_mline));
+						/* Start with media type, port and protocol */
+						char type[32];
+						char proto[64];
+						if(sscanf(line, "m=%31s %"SCNu16" %63s %*s", type, &m->port, proto) != 3) {
+							if(error)
+								g_snprintf(error, errlen, "Invalid m= line: %s", line);
+							success = FALSE;
+							break;
+						}
+						if(!strcasecmp(type, "audio"))
+							m->type = JANUS_SDP_AUDIO;
+						else if(!strcasecmp(type, "video"))
+							m->type = JANUS_SDP_VIDEO;
+						else if(!strcasecmp(type, "application"))
+							m->type = JANUS_SDP_APPLICATION;
+						else
+							m->type = JANUS_SDP_OTHER;
+						m->type_str = g_strdup(type);
+						m->proto = g_strdup(proto);
+						m->direction = JANUS_SDP_SENDRECV;
+						/* Now let's check the payload types */
+						gchar **mline_parts = g_strsplit(line+2, " ", -1);
+						if(!mline_parts) {
+							if(error)
+								g_snprintf(error, errlen, "Invalid m= line (no payload types): %s", line);
+							success = FALSE;
+							break;
+						}
+						int mindex = 0;
```

Comment:
not all protocols use a list of nubers in the m line: https://tools.ietf.org/html/rfc4566#section-5.14 maybe you should only extract those fopr RTP profiles and just keep a copy as a char* field called "fmt" for others? That may also help when recomposing SDPs if the ptypes are not changed (you could just copy m->fmt)